File: node-sdk/worker/worker.js
const WebSocket = require('ws');
const EventEmitter = require('events');
const crypto = require('crypto');

class Worker extends EventEmitter {
  constructor(config) {
    super();
    this.config = {
      schedulerURL: config.schedulerURL,
      workerGroup: config.workerGroup,
      maxRetry: config.maxRetry || 3,
      pingInterval: config.pingInterval || 30, // 秒
    };

    this.ws = null;
    this.methods = new Map();
    this.docs = new Map();
    this.running = false;
    this.reconnect = false;
    this.pingTimer = null;
    this.reconnectTimer = null;
  }

  /**
   * 注册方法
   * @param {string} name - 方法名
   * @param {Function} handler - 处理函数，签名: async (params) => result
   * @param {string[]} docs - 文档说明
   */
  registerMethod(name, handler, ...docs) {
    if (typeof handler !== 'function') {
      throw new Error('Handler must be a function');
    }

    this.methods.set(name, handler);
    this.docs.set(name, docs);
  }

  /**
   * 启动Worker
   */
  async start() {
    this.running = true;
    this.reconnect = true;

    await this.connect();
    this.startKeepAlive();

    console.log(`Worker ${this.config.workerGroup} started`);
  }

  /**
   * 连接到调度器
   */
  async connect() {
    return new Promise((resolve, reject) => {
      let retryCount = 0;

      const attemptConnect = () => {
        if (retryCount >= this.config.maxRetry) {
          reject(
            new Error(
              `Failed to connect after ${this.config.maxRetry} attempts`
            )
          );
          return;
        }

        try {
          this.ws = new WebSocket(this.config.schedulerURL);

          this.ws.on('open', () => {
            console.log('Connected to scheduler');

            // 发送注册信息
            const registration = {
              group: this.config.workerGroup,
              methods: this.getMethodsWithDocs(),
            };

            this.ws.send(JSON.stringify(registration));
            resolve();
          });

          this.ws.on('message', (data) => {
            this.handleMessage(data);
          });

          this.ws.on('close', (code, reason) => {
            console.log(`Connection closed: ${code} ${reason}`);
            this.ws = null;

            if (this.running && this.reconnect) {
              console.log('Attempting to reconnect in 5 seconds...');
              this.reconnectTimer = setTimeout(() => {
                this.connect().catch((err) => {
                  console.error('Reconnect failed:', err);
                });
              }, 5000);
            }
          });

          this.ws.on('error', (error) => {
            console.error('WebSocket error:', error);
            if (retryCount === 0) {
              retryCount++;
              setTimeout(attemptConnect, 1000 * retryCount);
            }
          });
        } catch {
          retryCount++;
          setTimeout(attemptConnect, 1000 * retryCount);
        }
      };

      attemptConnect();
    });
  }

  /**
   * 处理接收到的消息
   */
  async handleMessage(data) {
    try {
      const message = JSON.parse(data.toString());

      switch (message.type) {
        case 'task':
          this.handleTask(message.taskId, message.method, message.params);
          break;
        case 'encrypted_task':
          this.handleEncryptedTask(
            message.taskId,
            message.method,
            message.params,
            message.key,
            message.crypto
          );
          break;
        case 'ping':
          this.sendPong();
          break;
        default:
          console.log('Unknown message type:', message.type);
      }
    } catch (error) {
      console.error('Error handling message:', error);
    }
  }

  /**
   * 处理任务
   */
  async handleTask(taskId, method, params) {
    try {
      const handler = this.methods.get(method);
      if (!handler) {
        this.sendResult(taskId, null, new Error('Method not found'));
        return;
      }

      // 调用注册的方法
      const result = await handler(params);
      this.sendResult(taskId, result, null);
    } catch (error) {
      this.sendResult(taskId, null, error);
    }
  }

  /**
   * 发送结果到调度器
   */
  sendResult(taskId, result, error) {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      console.error(
        `Cannot send result for task ${taskId}: connection is not open`
      );
      return;
    }

    const response = {
      type: 'result',
      taskId: taskId,
    };

    if (error) {
      response.error = error.message;
    } else {
      response.result = result;
    }

    try {
      this.ws.send(JSON.stringify(response));
    } catch (err) {
      console.error(`Failed to send result for task ${taskId}:`, err);
    }
  }

  /**
   * 发送pong响应
   */
  sendPong() {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return;
    }

    try {
      this.ws.send(JSON.stringify({ type: 'pong' }));
    } catch (error) {
      console.error('Failed to send pong:', error);
    }
  }

  /**
   * 启动心跳保持
   */
  startKeepAlive() {
    this.pingTimer = setInterval(() => {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        try {
          this.ws.send(JSON.stringify({ type: 'ping' }));
        } catch (error) {
          console.error('Ping failed:', error);
        }
      }
    }, this.config.pingInterval * 1000);
  }

  /**
   * 停止Worker
   */
  stop() {
    if (!this.running) {
      return;
    }

    this.running = false;
    this.reconnect = false;

    // 清理定时器
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }

    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    // 关闭WebSocket连接
    if (this.ws) {
      this.ws.close(1000, 'Worker stopped');
      this.ws = null;
    }

    console.log('Worker stopped');
  }

  /**
   * 处理加密任务
   */
  async handleEncryptedTask(taskId, method, encryptedParams, key, cryptoSalt) {
    try {
      const handler = this.methods.get(method);
      if (!handler) {
        this.sendResult(taskId, null, new Error('Method not found'));
        return;
      }

      // 解密参数
      const decryptedParams = this.decryptData(encryptedParams, key, cryptoSalt);
      const params = JSON.parse(decryptedParams);

      // 调用注册的方法
      const result = await handler(params);

      // 加密结果
      const encryptedResult = this.encryptData(JSON.stringify(result), key, cryptoSalt);
      this.sendResult(taskId, encryptedResult, null);
    } catch (error) {
      this.sendResult(taskId, null, error);
    }
  }

  /**
   * AES-GCM加密数据
   */
  encryptData(data, saltedKey, cryptoSalt) {
    // 还原原始密钥
    const originalKey = this.unsaltKey(saltedKey, cryptoSalt);
    
    // 使用原始密钥生成AES密钥
    const hash = crypto.createHash('sha256').update(originalKey).digest();
    
    // 使用密钥生成确定性nonce（前12字节）
    const nonce = hash.slice(0, 12);
    
    // 创建cipher
    const cipher = crypto.createCipherGCM('aes-256-gcm');
    cipher.setIV(nonce);
    
    // 加密数据
    let encrypted = cipher.update(data, 'utf8', 'base64');
    encrypted += cipher.final('base64');
    
    // 获取认证标签
    const authTag = cipher.getAuthTag();
    
    // 合并加密数据和认证标签
    const combined = Buffer.concat([
      Buffer.from(encrypted, 'base64'),
      authTag
    ]);
    
    return combined.toString('base64');
  }

  /**
   * AES-GCM解密数据
   */
  decryptData(encryptedData, saltedKey, cryptoSalt) {
    // 还原原始密钥
    const originalKey = this.unsaltKey(saltedKey, cryptoSalt);
    
    // 解码base64
    const combined = Buffer.from(encryptedData, 'base64');
    
    // 分离加密数据和认证标签（最后16字节是认证标签）
    const encrypted = combined.slice(0, -16);
    const authTag = combined.slice(-16);
    
    // 使用原始密钥生成AES密钥
    const hash = crypto.createHash('sha256').update(originalKey).digest();
    
    // 使用密钥生成确定性nonce（前12字节）
    const nonce = hash.slice(0, 12);
    
    // 创建decipher
    const decipher = crypto.createDecipherGCM('aes-256-gcm');
    decipher.setIV(nonce);
    decipher.setAuthTag(authTag);
    
    // 解密数据
    let decrypted = decipher.update(encrypted, null, 'utf8');
    decrypted += decipher.final('utf8');
    
    return decrypted;
  }

  /**
   * 还原加盐密钥
   */
  unsaltKey(saltedKey, cryptoSalt) {
    // 解码base64
    const combined = Buffer.from(saltedKey, 'base64');
    
    // 分离加密数据和认证标签（最后16字节是认证标签）
    const encrypted = combined.slice(0, -16);
    const authTag = combined.slice(-16);
    
    // 使用crypto作为盐值生成AES密钥
    const hash = crypto.createHash('sha256').update(cryptoSalt).digest();
    
    // 使用盐值生成确定性nonce（前12字节）
    const nonce = hash.slice(0, 12);
    
    // 创建decipher
    const decipher = crypto.createDecipherGCM('aes-256-gcm');
    decipher.setIV(nonce);
    decipher.setAuthTag(authTag);
    
    // 解密密钥
    let decrypted = decipher.update(encrypted, null, 'utf8');
    decrypted += decipher.final('utf8');
    
    return decrypted;
  }

  /**
   * 获取方法和文档信息
   */
  getMethodsWithDocs() {
    const methods = [];
    for (const [name] of this.methods) {
      methods.push({
        name: name,
        docs: this.docs.get(name) || [],
      });
    }
    return methods;
  }
}

module.exports = Worker;

File: rust-sdk/src/worker/worker.rs
//! Worker implementation for handling distributed tasks

use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::{mpsc, RwLock};
use tokio::time::{interval, sleep, timeout};
use tokio_tungstenite::{connect_async, tungstenite::Message};
use url::Url;
use crate::crypto::{encrypt_data, decrypt_data, unsalt_key};
use crate::error::{Result, SdkError};
use super::{TASK_STATUS_ERROR};

/// Worker configuration
#[derive(Debug, Clone)]
pub struct Config {
    /// Scheduler WebSocket URL
    pub scheduler_url: String,
    /// Worker group name
    pub worker_group: String,
    /// Maximum connection retry attempts
    pub max_retry: usize,
    /// Ping interval in seconds
    pub ping_interval: u64,
}

/// Method handler function type
pub type MethodHandler = Box<dyn Fn(Value) -> Result<Value> + Send + Sync>;

/// Method definition with handler and documentation
#[derive(Clone)]
pub struct Method {
    pub name: String,
    pub handler: Arc<MethodHandler>,
    pub docs: Vec<String>,
}

/// WebSocket message types
#[derive(Deserialize, Debug)]
#[serde(tag = "type")]
enum IncomingMessage {
    #[serde(rename = "task")]
    Task {
        #[serde(rename = "taskId")]
        task_id: String,
        method: String,
        params: Value,
    },
    #[serde(rename = "encrypted_task")]
    EncryptedTask {
        #[serde(rename = "taskId")]
        task_id: String,
        method: String,
        params: Value,
        key: String,
        crypto: String,
    },
    #[serde(rename = "ping")]
    Ping,
}

#[derive(Serialize, Debug)]
#[serde(tag = "type")]
enum OutgoingMessage {
    #[serde(rename = "result")]
    Result {
        #[serde(rename = "taskId")]
        task_id: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        result: Option<Value>,
        #[serde(skip_serializing_if = "Option::is_none")]
        error: Option<String>,
    },
    #[serde(rename = "pong")]
    Pong,
}

#[derive(Serialize, Debug)]
struct RegistrationMessage {
    group: String,
    methods: Vec<MethodInfo>,
}

#[derive(Serialize, Debug)]
struct MethodInfo {
    name: String,
    docs: Vec<String>,
}

/// Distributed task worker
pub struct Worker {
    config: Config,
    methods: Arc<RwLock<HashMap<String, Method>>>,
    running: Arc<RwLock<bool>>,
    shutdown_tx: Option<mpsc::Sender<()>>,
}

impl Worker {
    /// Creates a new worker with the given configuration
    /// 
    /// # Arguments
    /// 
    /// * `config` - Worker configuration
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use go_server_rust_sdk::worker::{Worker, Config};
    /// 
    /// let config = Config {
    ///     scheduler_url: "ws://localhost:8080/api/worker/connect/123456".to_string(),
    ///     worker_group: "math".to_string(),
    ///     max_retry: 5,
    ///     ping_interval: 5,
    /// };
    /// 
    /// let worker = Worker::new(config);
    /// ```
    pub fn new(config: Config) -> Self {
        Self {
            config,
            methods: Arc::new(RwLock::new(HashMap::new())),
            running: Arc::new(RwLock::new(false)),
            shutdown_tx: None,
        }
    }

    /// Registers a method handler with the worker
    /// 
    /// # Arguments
    /// 
    /// * `name` - Method name
    /// * `handler` - Function that handles the method call
    /// * `docs` - Documentation strings for the method
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use go_server_rust_sdk::worker::{Worker, Config};
    /// use serde_json::{json, Value};
    /// 
    /// # let config = Config {
    /// #     scheduler_url: "ws://localhost:8080/api/worker/connect/123456".to_string(),
    /// #     worker_group: "math".to_string(),
    /// #     max_retry: 5,
    /// #     ping_interval: 5,
    /// # };
    /// let mut worker = Worker::new(config);
    /// 
    /// worker.register_method("add", |params: Value| {
    ///     let a = params["a"].as_f64().unwrap_or(0.0);
    ///     let b = params["b"].as_f64().unwrap_or(0.0);
    ///     Ok(json!(a + b))
    /// }, vec!["Add two numbers".to_string()]);
    /// ```
    pub fn register_method<F>(&mut self, name: impl Into<String>, handler: F, docs: Vec<String>)
    where
        F: Fn(Value) -> Result<Value> + Send + Sync + 'static,
    {
        let method = Method {
            name: name.into(),
            handler: Arc::new(Box::new(handler)),
            docs,
        };
        
        // We need to use a blocking approach here since this is a sync method
        let methods = self.methods.clone();
        tokio::spawn(async move {
            let mut methods_guard = methods.write().await;
            methods_guard.insert(method.name.clone(), method);
        });
    }

    /// Starts the worker with automatic reconnection support
    /// 
    /// This method will block until the worker is stopped.
    /// 
    /// # Returns
    /// 
    /// A `Result` indicating success or failure
    /// 
    /// # Example
    /// 
    /// ```rust,no_run
    /// use go_server_rust_sdk::worker::{Worker, Config};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// # let config = Config {
    /// #     scheduler_url: "ws://localhost:8080/api/worker/connect/123456".to_string(),
    /// #     worker_group: "math".to_string(),
    /// #     max_retry: 5,
    /// #     ping_interval: 5,
    /// # };
    /// let mut worker = Worker::new(config);
    /// // Register methods...
    /// worker.start().await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn start(&mut self) -> Result<()> {
        *self.running.write().await = true;
        
        let (shutdown_tx, mut shutdown_rx) = mpsc::channel::<()>(1);
        self.shutdown_tx = Some(shutdown_tx);

        log::info!("Worker {} starting", self.config.worker_group);

        loop {
            // Check if we should stop
            if !*self.running.read().await {
                break;
            }

            // Try to connect and run
            match self.connect_and_run(&mut shutdown_rx).await {
                Ok(_) => {
                    log::info!("Worker connection closed normally");
                    break;
                }
                Err(e) => {
                    log::error!("Worker connection failed: {}", e);
                    if *self.running.read().await {
                        log::info!("Retrying connection in 5 seconds...");
                        sleep(Duration::from_secs(5)).await;
                    }
                }
            }
        }

        log::info!("Worker {} stopped", self.config.worker_group);
        Ok(())
    }

    /// Stops the worker
    /// 
    /// This method will signal the worker to stop and close all connections.
    /// 
    /// # Example
    /// 
    /// ```rust,no_run
    /// use go_server_rust_sdk::worker::{Worker, Config};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// # let config = Config {
    /// #     scheduler_url: "ws://localhost:8080/api/worker/connect/123456".to_string(),
    /// #     worker_group: "math".to_string(),
    /// #     max_retry: 5,
    /// #     ping_interval: 5,
    /// # };
    /// let mut worker = Worker::new(config);
    /// 
    /// // In another task or signal handler:
    /// worker.stop().await;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn stop(&mut self) {
        *self.running.write().await = false;
        
        if let Some(tx) = &self.shutdown_tx {
            let _ = tx.send(()).await;
        }
    }

    async fn connect_and_run(&self, shutdown_rx: &mut mpsc::Receiver<()>) -> Result<()> {
        let url = Url::parse(&self.config.scheduler_url)?;
        let (ws_stream, _) = connect_async(url).await?;
        let (mut ws_sender, mut ws_receiver) = ws_stream.split();

        // Send registration message
        let methods = self.get_methods_info().await;
        let registration = RegistrationMessage {
            group: self.config.worker_group.clone(),
            methods,
        };
        
        let registration_msg = serde_json::to_string(&registration)?;
        ws_sender.send(Message::Text(registration_msg)).await?;

        log::info!("Worker {} connected and registered", self.config.worker_group);

        // Start ping interval
        let mut ping_interval = interval(Duration::from_secs(self.config.ping_interval));
        
        loop {
            tokio::select! {
                // Handle shutdown signal
                _ = shutdown_rx.recv() => {
                    log::info!("Received shutdown signal");
                    let _ = ws_sender.close().await;
                    break;
                }
                
                // Handle ping interval
                _ = ping_interval.tick() => {
                    let ping_msg = serde_json::to_string(&OutgoingMessage::Pong)?;
                    if let Err(e) = ws_sender.send(Message::Text(ping_msg)).await {
                        log::error!("Failed to send ping: {}", e);
                        break;
                    }
                }
                
                // Handle incoming messages
                msg = ws_receiver.next() => {
                    match msg {
                        Some(Ok(Message::Text(text))) => {
                            if let Err(e) = self.handle_message(&text, &mut ws_sender).await {
                                log::error!("Error handling message: {}", e);
                            }
                        }
                        Some(Ok(Message::Close(_))) => {
                            log::info!("WebSocket connection closed by server");
                            break;
                        }
                        Some(Err(e)) => {
                            log::error!("WebSocket error: {}", e);
                            break;
                        }
                        None => {
                            log::info!("WebSocket stream ended");
                            break;
                        }
                        _ => {}
                    }
                }
            }
        }

        Ok(())
    }

    async fn handle_message(
        &self,
        text: &str,
        ws_sender: &mut futures_util::stream::SplitSink<
            tokio_tungstenite::WebSocketStream<tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>>,
            Message,
        >,
    ) -> Result<()> {
        let message: IncomingMessage = serde_json::from_str(text)?;

        match message {
            IncomingMessage::Task { task_id, method, params } => {
                self.handle_task(task_id, method, params, ws_sender).await?
            }
            IncomingMessage::EncryptedTask { task_id, method, params, key, crypto } => {
                self.handle_encrypted_task(task_id, method, params, key, crypto, ws_sender).await?
            }
            IncomingMessage::Ping => {
                let pong_msg = serde_json::to_string(&OutgoingMessage::Pong)?;
                ws_sender.send(Message::Text(pong_msg)).await?;
            }
        }

        Ok(())
    }

    async fn handle_task(
        &self,
        task_id: String,
        method: String,
        params: Value,
        ws_sender: &mut futures_util::stream::SplitSink<
            tokio_tungstenite::WebSocketStream<tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>>,
            Message,
        >,
    ) -> Result<()> {
        let methods = self.methods.read().await;
        let method_handler = methods.get(&method).cloned();
        drop(methods);

        let response = match method_handler {
            Some(handler) => {
                match (handler.handler)(params) {
                    Ok(result) => OutgoingMessage::Result {
                        task_id,
                        result: Some(result),
                        error: None,
                    },
                    Err(e) => OutgoingMessage::Result {
                        task_id,
                        result: None,
                        error: Some(e.to_string()),
                    },
                }
            }
            None => OutgoingMessage::Result {
                task_id,
                result: None,
                error: Some(format!("Method '{}' not found", method)),
            },
        };

        let response_text = serde_json::to_string(&response)?;
        ws_sender.send(Message::Text(response_text)).await?;

        Ok(())
    }

    async fn handle_encrypted_task(
        &self,
        task_id: String,
        method: String,
        encrypted_params: Value,
        salted_key: String,
        crypto: String,
        ws_sender: &mut futures_util::stream::SplitSink<
            tokio_tungstenite::WebSocketStream<tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>>,
            Message,
        >,
    ) -> Result<()> {
        let methods = self.methods.read().await;
        let method_handler = methods.get(&method).cloned();
        drop(methods);

        let response = match method_handler {
            Some(handler) => {
                // Decrypt parameters
                match self.decrypt_task_params(encrypted_params, &salted_key, &crypto).await {
                    Ok(params) => {
                        // Execute method
                        match (handler.handler)(params) {
                            Ok(result) => {
                                // Encrypt result
                                match self.encrypt_task_result(result, &salted_key, &crypto).await {
                                    Ok(encrypted_result) => OutgoingMessage::Result {
                                        task_id,
                                        result: Some(encrypted_result),
                                        error: None,
                                    },
                                    Err(e) => OutgoingMessage::Result {
                                        task_id,
                                        result: None,
                                        error: Some(format!("Failed to encrypt result: {}", e)),
                                    },
                                }
                            }
                            Err(e) => OutgoingMessage::Result {
                                task_id,
                                result: None,
                                error: Some(e.to_string()),
                            },
                        }
                    }
                    Err(e) => OutgoingMessage::Result {
                        task_id,
                        result: None,
                        error: Some(format!("Failed to decrypt params: {}", e)),
                    },
                }
            }
            None => OutgoingMessage::Result {
                task_id,
                result: None,
                error: Some(format!("Method '{}' not found", method)),
            },
        };

        let response_text = serde_json::to_string(&response)?;
        ws_sender.send(Message::Text(response_text)).await?;

        Ok(())
    }

    async fn decrypt_task_params(
        &self,
        encrypted_params: Value,
        salted_key: &str,
        crypto: &str,
    ) -> Result<Value> {
        // Extract encrypted string from JSON
        let encrypted_str = encrypted_params
            .as_str()
            .ok_or_else(|| SdkError::Crypto("Invalid encrypted params format".to_string()))?;

        // Parse salt from crypto string
        let salt: i32 = crypto.parse()
            .map_err(|_| SdkError::Crypto("Invalid crypto salt format".to_string()))?;

        // Unsalt the key
        let original_key = unsalt_key(salted_key, salt)?;

        // Decrypt the parameters
        decrypt_data(encrypted_str, &original_key)
    }

    async fn encrypt_task_result(
        &self,
        result: Value,
        salted_key: &str,
        crypto: &str,
    ) -> Result<Value> {
        // Parse salt from crypto string
        let salt: i32 = crypto.parse()
            .map_err(|_| SdkError::Crypto("Invalid crypto salt format".to_string()))?;

        // Unsalt the key
        let original_key = unsalt_key(salted_key, salt)?;

        // Serialize result to JSON string
        let result_str = serde_json::to_string(&result)?;

        // Encrypt the result using the original key
        let encrypted_result = encrypt_data(&Value::String(result_str), &original_key)?;

        Ok(Value::String(encrypted_result))
    }

    async fn get_methods_info(&self) -> Vec<MethodInfo> {
        let methods = self.methods.read().await;
        methods
            .values()
            .map(|method| MethodInfo {
                name: method.name.clone(),
                docs: method.docs.clone(),
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_worker_creation() {
        let config = Config {
            scheduler_url: "ws://localhost:8080/api/worker/connect/123456".to_string(),
            worker_group: "test".to_string(),
            max_retry: 3,
            ping_interval: 5,
        };

        let worker = Worker::new(config.clone());
        assert_eq!(worker.config.worker_group, "test");
        assert_eq!(worker.config.max_retry, 3);
    }

    #[tokio::test]
    async fn test_method_registration() {
        let config = Config {
            scheduler_url: "ws://localhost:8080/api/worker/connect/123456".to_string(),
            worker_group: "test".to_string(),
            max_retry: 3,
            ping_interval: 5,
        };

        let mut worker = Worker::new(config);
        
        worker.register_method("test_method", |params: Value| {
            Ok(json!({"received": params}))
        }, vec!["Test method".to_string()]);

        // Give some time for the async registration to complete
        tokio::time::sleep(Duration::from_millis(10)).await;

        let methods = worker.methods.read().await;
        assert!(methods.contains_key("test_method"));
    }
}